core/src/main/java/com/linecorp/armeria/server/jsonrpc/JsonRpcService.java

```diff
@@ -0,0 +1,186 @@
+import com.linecorp.armeria.common.jsonrpc.JsonRpcResponse;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A JSON-RPC {@link HttpService}.
+ */
+@UnstableApi
+public final class JsonRpcService implements HttpService {
+    private static final ObjectMapper mapper = JacksonUtil.newDefaultObjectMapper();
+
+    private final Map<String, JsonRpcHandler> methodHandlers;
+
+    /**
+    * Returns a new {@link JsonRpcServiceBuilder}.
+    */
+    public static JsonRpcServiceBuilder builder() {
+        return new JsonRpcServiceBuilder();
+    }
+
+    JsonRpcService(Map<String, JsonRpcHandler> methodHandlers) {
+        this.methodHandlers = requireNonNull(methodHandlers, "methodHandlers");
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) {
+        if (req.method() != HttpMethod.POST) {
+            return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+
+        if (req.contentType() == null || !req.contentType().isJson()) {
+            return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE);
+        }
+
+        return HttpResponse.of(
+            req.aggregate()
+               .thenApply(JsonRpcService::parseRequestContentAsJson)
+               .thenApply(json -> dispatchRequest(ctx, json))
+               .exceptionally(e -> {
+                    if (e.getCause() instanceof IllegalArgumentException) {
+                        return HttpResponse.ofJson(HttpStatus.BAD_REQUEST, JsonRpcError.PARSE_ERROR);
+                    }
+                    return HttpResponse.ofJson(HttpStatus.BAD_REQUEST,
+                        JsonRpcError.INTERNAL_ERROR.withData(e.getMessage()));
+                })
+            );
+    }
+
+    @VisibleForTesting
+    static JsonNode parseRequestContentAsJson(AggregatedHttpRequest request) {
+        try {
+            return mapper.readTree(request.contentUtf8());
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    @VisibleForTesting
+    HttpResponse dispatchRequest(ServiceRequestContext ctx, JsonNode rawRequest) {
+        if (rawRequest.isObject()) {
+            return handleUnaryRequest(ctx, rawRequest);
+        } else {
+            return HttpResponse.ofJson(HttpStatus.BAD_REQUEST, JsonRpcError.PARSE_ERROR);
+        }
+    }
+
+    private HttpResponse handleUnaryRequest(ServiceRequestContext ctx, JsonNode unary) {
+        return HttpResponse.of(
+                executeRpcCall(ctx, unary)
+                    .thenApply(this::toHttpResponse));
+    }
+
+    private HttpResponse toHttpResponse(DefaultJsonRpcResponse response) {
+        if (response == null) {
+            return HttpResponse.of(HttpStatus.ACCEPTED, MediaType.PLAIN_TEXT_UTF_8, "");
```

Comment:
We may not need a content-type and an empty string for accepted status.
```suggestion
            return HttpResponse.of(HttpStatus.ACCEPTED);
```

Reply:
Understood. We could make the response body empty by specifying response headers only.
```java
HttpResponse.of(ResponseHeaders.of(HttpStatus.ACCEPTED));
```